src/main/java/com/github/fridujo/rabbitmq/mock/MockQueue.java

```diff
@@ -82,17 +81,26 @@ private boolean deliverToConsumerIfPossible() {
 
         delivered = deadLetterFirstMessageIfExpired();
 
-        if (consumersByTag.size() > 0) {
+        if (hasConsumers()) {
             Message message = messages.poll();
             if (message != null) {
                 if (message.isExpired()) {
                     deadLetterWithReason(message, DeadLettering.ReasonType.EXPIRED);
                 } else {
-                    int index = consumerRollingSequence.incrementAndGet() % consumersByTag.size();
-                    ConsumerAndTag nextConsumer = new ArrayList<>(consumersByTag.values()).get(index);
+                    List<ConsumerAndTag> consumerAndTags;
+
+                    synchronized(consumersByTag) {
```

Comment:
I am not confortable with `synchronized`blocks, it makes reading the code more difficult.

Could we not switch to a `ConcurrentHashMap` and avoid these blocks ?

Reply:
You used `LinkedHashMap`s here in the original code, if these could have been regular `HashMap`s then we could switch to `ConcurrentHashMap` (there is no `ConcurrentLinkedHashMap`).

Reply:
I think `unackedMessagesByDeliveryTag` can't be changed to a non-linked hash map without changing the order of messages on the queue it represents.

Reply:
Yes, my bad, you are completely right.

However for this purpose there is still `Collections#synchronyzedMap(new LinkedHashMap)`.
There is still the `synchronized` penalty (but the kind of use of this project, it seems acceptable), but it does favor a more readable code

WDYT ?

Reply:
I can wrap it with `synchronizedMap`, which would avoid some of the synchronized blocks, but I don't think all of them can be avoided, for example:

        synchronized(unackedMessagesByDeliveryTag) {
            Message message = unackedMessagesByDeliveryTag.remove(deliveryTag);

            unackedDeliveryTagsByConsumerTag.forEach((ctag, deliveryTags) -> deliveryTags.remove(deliveryTag));

            return message;
        }

Here it looks important to modify these two maps at the same time (which is why I synchronized on `unackedMessagesByDeliveryTag` for both these maps, instead of two separate ones).

Maybe there is one or two more where I have to use `synchronized`, but I think the others could be converted.  I would have to use the `forEach` function in a few places to iterate over the maps while holding the lock (instead of the `for` statement).

The performance shouldn't be much of an issue relative to what the rest of the code is doing, and I think correctness always trumps performance.

Reply:
I agree

Reply:
I've updated the code.

I should note that I doubt all problems are fixed in this code when it comes to concurrency. There are three maps in play and the priority queue.  Synchronization and concurrent version only ensures that these structures are internally consistent. A better look would be needed to see if some of these structures should be consistent as a group.

For example, let's say you have two maps that should always have the same keys:

        map1.put("x", "12345");  // synchronized maps
        map2.put("x", "abcde");

Another thread might see `map1` containing "x" while `map2` doesn't (yet) contain "x".  Both maps are internally consistent, but another thread might iterate all keys from `map1`, then get the same key from `map2` and may be surprised to see that it is sometimes `null`:

        map1.keySet().forEach(k -> {
             map2.get(k);  // may return null !
        });

To avoid this, the code above should have been written as:

        synchronized(lockForBothMaps) {
           map1.put("x", "12345");
           map2.put("x", "abcde");
        }

And:

        synchronized(lockForBothMaps) {
            map1.keySet().forEach(k -> {
                 map2.get(k);  // may return null !
            });
        }

I also must say that using synchronized maps makes the code a bit less clear; one could be tempted to change something like:

                    List<ConsumerAndTag> consumerAndTags = new ArrayList<>(consumersByTag.size());

                    consumersByTag.values().forEach(consumerAndTags::add);  // iterates while synchronized

To:

                    List<ConsumerAndTag> consumerAndTags = new ArrayList<>(consumerAndTags.values());

But this is not equivalent as the iteration logic in `ArrayList` will NOT properly lock the `consumerAndTags` map.

I hope this is statisfactory for now... we really could use a release with these changes.